src/controller/src/clusters.rs

```diff
@@ -280,32 +280,54 @@ where
                 }
             })
             // TODO(guswynn): make this configurable.
-            .buffer_unordered(50)
-            // `try_collect` and `collect` are the only safe ways to process a
-            // `buffer_unordered`. See the docs in `mz_storage_client::controller` for more
-            // details.
-            .try_collect()
-            .await?;
-
-        for (
-            cluster_id,
-            replica_id,
-            compute_config,
-            storage_location,
-            compute_location,
-            metrics_task_join_handle,
-        ) in replicas
-        {
-            if let Some(jh) = metrics_task_join_handle {
-                self.metrics_tasks.insert(replica_id, jh);
-            }
-            self.storage.connect_replica(cluster_id, storage_location);
-            self.active_compute().add_replica_to_instance(
+            .buffer_unordered(50);
+
+        // _Usually_ `try_collect` and `collect` are the only safe ways to process a
+        // `buffer_unordered`, but if we ensure we don't do
+        // any async work in this loop, we are fine.
+        //
+        // See the docs in `mz_storage_client::controller` for more
+        let mut replicas: BTreeMap<_, Vec<_>> = BTreeMap::new();
+        while let Some(res) = replica_stream.next().await {
+            let (
                 cluster_id,
                 replica_id,
+                compute_config,
+                storage_location,
                 compute_location,
+                metrics_task_join_handle,
+            ) = res?;
+
+            replicas.entry(cluster_id).or_default().push((
+                replica_id,
                 compute_config,
-            )?;
+                storage_location,
+                compute_location,
+                metrics_task_join_handle,
+            ));
+        }
+        drop(replica_stream);
+
+        for (cluster_id, replicas) in replicas {
+            // We only connect to the last replica (chosen arbitrarily)
+            // for storage, until we support multi-replica storage objects
+            self.storage
+                // the `.2` is very specifically for the `storage_location`,
+                // not the `compute_location`
```

Comment:
This is very error-prone; I think the objects in `replicas` could use a more legit type than a long tuple. Same goes for the stuff in `replica_stream`.

Reply:
Strong agree. Looks like the `replicas` arg to this method was also changed to be a vector of 4-tuples, and that should also have a proper structed extracted.

Reply:
good point! will change to a named struct

Reply:
updated with this!